typing/typecore.ml

```diff
@@ -4106,27 +4268,56 @@ and type_expect_
         raise (Typetexp.Error (loc, env,
           Unsupported_extension Polymorphic_parameters));
       type_function ?in_function loc sexp.pexp_attributes env
-                    expected_mode ty_expected_explained l ~has_local
+                    expected_mode ty_expected_explained l ~amodes
                     ~has_poly [Ast_helper.Exp.case spat sbody]
   | Pexp_function caselist ->
+      let amodes = Mode.Alloc.newvar () in
       type_function ?in_function
         loc sexp.pexp_attributes env expected_mode
-        ty_expected_explained Nolabel ~has_local:false ~has_poly:false caselist
+        ty_expected_explained Nolabel ~amodes ~has_poly:false caselist
+  | Pexp_apply
+      ({ pexp_desc = Pexp_extension({txt = ("ocaml.unique" | "unique" | "extension.unique" as txt)}, PStr []) },
+       [Nolabel, sbody]) ->
+      if txt = "extension.unique" && not (Language_extension.is_enabled Unique) then
+        raise (Typetexp.Error (loc, Env.empty, Unsupported_extension Unique));
+      let mode = if mode_cross env ty_expected then mode_max
+      else begin
+        let mode = mode_unique expected_mode in
+        submode ~loc ~env ~reason:Other mode.mode expected_mode;
+        mode
+      end
+      in
+      let exp = type_expect ?in_function ~recarg env mode sbody ty_expected_explained in
+      { exp with exp_loc = loc }
+  | Pexp_apply
+      ({ pexp_desc = Pexp_extension({txt = ("ocaml.once" | "once" | "extension.once" as txt)}, PStr []) },
+      [Nolabel, sbody]) ->
+      if txt = "extension.once" && not (Language_extension.is_enabled Unique) then
+        raise (Typetexp.Error (loc, Env.empty, Unsupported_extension Unique));
+      let mode = if mode_cross env ty_expected then mode_max
+      else begin
+        let mode = mode_once expected_mode in
+        submode ~loc ~env ~reason:Other mode.mode expected_mode;
+        mode
+      end
+      in
+      let exp = type_expect ?in_function ~recarg env mode sbody ty_expected_explained in
+      { exp with exp_loc = loc }
```

Comment:
I'm unsure about this - when we write `once_ exp` do we mean that `exp` needs to be strictly `once`? Or is `many` also ok?

Reply:
`once_ exp` does not make any requirement on `exp` -- it expects it to produce a `once` value but that's the weakest mode so doesn't change anything.